broker/service/spreadsheet_upload_service.py

```diff
@@ -16,28 +18,55 @@ def __init__(self, ingest_api: IngestApi, storage_service: SpreadsheetStorageSer
         self.storage_service = storage_service
         self.importer = importer
 
-    def async_upload(self, token, request_file, is_update, project_uuid=None):
-        if token is None:
-            raise SpreadsheetUploadError(401, "An authentication token must be supplied when uploading a spreadsheet")
-
-        self.ingest_api.set_token(token)
-        submission_resource = self.ingest_api.create_submission(update_submission=is_update)
+    def async_upload(self, token, request_file, is_update, project_uuid=None, submission_uuid=None):
+        self._assert_token_exists(token)
+        submission_resource = self._create_or_get_submission(submission_uuid)
 
+        submission_uuid = submission_resource["uuid"]["uuid"]
+        submission_url = submission_resource["_links"]["self"]["href"]
         filename = secure_filename(request_file.filename)
 
-        submission_uuid = submission_resource["uuid"]["uuid"]
-        path = self.storage_service.store(submission_uuid, filename, request_file.read())
+        if is_update:
+            path = self._store_spreadsheet_updates(filename, request_file, submission_uuid)
+            thread = threading.Thread(target=self.upload_updates, args=(submission_url, path))
+        else:
+            path = self.storage_service.store_submission_spreadsheet(submission_uuid, filename, request_file.read())
+            thread = threading.Thread(target=self.upload, args=(submission_url, path, project_uuid))
 
-        thread = threading.Thread(target=self._upload,
-                                  args=(submission_resource, path, is_update, project_uuid))
         thread.start()
+
         return submission_resource
 
-    def _upload(self, submission_resource, path, is_update, project_uuid=None):
+    def _store_spreadsheet_updates(self, filename, request_file, submission_uuid):
+        submission_directory = self.storage_service.get_submission_dir(submission_uuid)
+        timestamp = time.strftime("%Y%m%d-%H%M%S")
+        filename_with_timestamp = f'{timestamp}_{filename}'
+        # TODO This spreadsheet containing the updates is not downloadable anywhere yet
+        path = self.storage_service.store_binary_file(submission_directory, filename_with_timestamp,
+                                                      request_file.read())
+        return path
+
+    def _create_or_get_submission(self, submission_uuid):
+        if submission_uuid:
+            submission_resource = self.ingest_api.get_submission_by_uuid(submission_uuid)
+        else:
+            submission_resource = self.ingest_api.create_submission()
+        return submission_resource
+
+    def _assert_token_exists(self, token):
+        if token is None:
+            raise SpreadsheetUploadError(401, "An authentication token must be supplied when uploading a spreadsheet")
+        self.ingest_api.set_token(token)
+
+    def upload(self, submission_url, path, project_uuid=None):
```

Comment:
Is `upload` for a full project rather than just updates?

Reply:
The upload function is for uploading spreadsheets with new entities. Sometime's the project is created before submission so we supply it if it already exists.